### PR TITLE
Added CosFace Loss

### PR DESCRIPTION
**Pull request by crsdvaibhav**
- Added CosFace Loss, with tests and documentation.
- Moved `l2_norm` from ArcFace Loss to `quaterion.utils` and updated the imports and `__init__.py`

### CODE REVIEW

**Comment by crsdvaibhav** [comment]
@monatis All checks have passed. Please could you review it? Thanks!

File: tests/eval/losses/test_cos_face_loss.py [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+            [1.0, 0.0, 0.9],
+            [1.2, -1.2, 0.01],
+            [-0.7, 0.0, 1.5],
+        ]
+    )
+
+    groups = torch.LongTensor([1, 2, 3, 3, 2, 1])
+
+    def test_batch_all(self):
+        loss = CosFaceLoss()
```
Comment (monatis) [inline tests/eval/losses/test_cos_face_loss.py:21]: You do not pass the required arguments to the loss in tests.

```suggestion
        loss = CosFaceLoss(self.embeddings.size()[1], 3)
```

**Review by monatis — CHANGES_REQUESTED** [review]

**Comment by crsdvaibhav** [comment]
@monatis Sorry, I must have missed changing it after reverting my commits. Thank you! Is there anything more before merging?

File: quaterion/loss/cos_face_loss.py [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+
+        """Compute loss value
+        Args:
+            embeddings: shape: (batch_size, vector_length) - Output embeddings from the
+                encoder.
+            groups: shape: (batch_size,) - Group ids associated with embeddings.
+        Returns:
+            Tensor: loss value.
+        """
+
```
Comment (monatis) [inline quaterion/loss/cos_face_loss.py:48]: Let's make a proper assertion.

```suggestion
        assert groups.ge(0).all() and groups.lt(self.kernel.size(1)).all(), \
            f"Invalid group ids: all the values must be between 0 (inclusive) and num_groups (exclusive), but given:  {groups}"

```

**Comment by crsdvaibhav** [comment]
@monatis Ran the tests locally. Made the requested changes. Anything more, to be done? Also, thankyou for the help!
test_cos_face_loss

**Comment by monatis** [comment]
@crsdvaibhav ArcFace has the assertion missing, and CosFace has it twice. See also #203 (thanks @KarahanS)

**Review by monatis — APPROVED** [review]

**Comment by crsdvaibhav** [comment]
> @crsdvaibhav ArcFace has the assertion missing, and CosFace has it twice. See also #203 (thanks @KarahanS)

I changed it. Thanks for the merge! I will move to #203
